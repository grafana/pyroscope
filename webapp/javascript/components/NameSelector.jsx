import React from "react";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { addLabel, receiveJSON } from "../redux/actions";

function NameSelector(props) {
  const { actions, names, labels } = props;
  const selectName = (event) => {
    actions.addLabel("__name__", event.target.value);
  };

  let selectedName = labels.filter((x) => x.name === "__name__")[0];
  selectedName = selectedName ? selectedName.value : "none";
  return (
    <span>
      Application:&nbsp;
      <select
        className="label-select"
        value={selectedName}
        onChange={selectName}
      >
        <option disabled key="Select an app..." value="Select an app...">
          Select an app...
        </option>
        {names &&
          names.map((name) => (
            <option key={name} value={name}>
              {name}
            </option>
          ))}
      </select>
    </span>
  );
}

const mapStateToProps = (state) => ({
  ...state,
});

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(
    {
      receiveJSON,
      addLabel,
    },
    dispatch
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(NameSelector);
