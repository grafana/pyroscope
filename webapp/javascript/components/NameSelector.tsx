// TODO reenable spreading lint
/* eslint-disable react/jsx-props-no-spreading */
import React, { useState, useEffect } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { useAppSelector, useAppDispatch } from '@pyroscope/redux/hooks';
import { appNameToQuery, queryToAppName } from '@utils/query';
import {
  selectAppNames,
  selectAppNamesState,
  reloadAppNames,
} from '@pyroscope/redux/reducers/newRoot';
import LoadingSpinner from '@ui/LoadingSpinner';
import Button from '@ui/Button';
import { faSyncAlt } from '@fortawesome/free-solid-svg-icons/faSyncAlt';
import Dropdown, { MenuItem, FocusableItem } from '@ui/Dropdown';
import { setQuery } from '../redux/actions';
import styles from './NameSelector.module.scss';

const defKey = 'Select an app...';

function NameSelector(props) {
  const { actions, names, query } = props;
  const appNamesState = useAppSelector(selectAppNamesState);
  const appNames = useAppSelector(selectAppNames);

  const [filter, setFilter] = useState('');

  const selectAppName = (name: string) => {
    const query = appNameToQuery(name);
    actions.setQuery(query);
  };

  const dispatch = useAppDispatch();

  // if there's no query set
  // set the first app as the default (if exists)
  useEffect(() => {
    if (!query) {
      const first = appNames[0];
      if (first) {
        const query = appNameToQuery(first);
        actions.setQuery(query);
      }
    }
  }, [query]);

  const selectedValue = queryToAppName(query).mapOr('', (q) => {
    if (appNames.indexOf(q) !== -1) {
      return q;
    }
    return '';
  });

  const filterOptions = (n: string) => {
    const f = filter.trim().toLowerCase();
    return n.toLowerCase().includes(f);
  };

  // TODO figure out this any
  const options = appNames.filter(filterOptions).map((name) => (
    <MenuItem
      key={name}
      value={name}
      onClick={() => selectAppName(name)}
      className={selectedValue === name ? 'active' : ''}
    >
      {name}
    </MenuItem>
  )) as any;

  const noApp =
    appNames.length > 0 ? null : <MenuItem>No App available</MenuItem>;

  return (
    <div className={styles.container}>
      Application:&nbsp;
      <Dropdown
        label="Select application"
        data-testid="app-name-selector"
        value={selectedValue}
        overflow="auto"
        position="anchor"
        menuButtonClassName={styles.menuButton}
      >
        {noApp}
        <FocusableItem>
          {({ ref }) => (
            <input
              ref={ref}
              type="text"
              placeholder="Type an app"
              value={filter}
              onChange={(e) => setFilter(e.target.value)}
            />
          )}
        </FocusableItem>

        {options}
      </Dropdown>
      <Button
        aria-label="Refresh Apps"
        icon={faSyncAlt}
        onClick={() => {
          dispatch(reloadAppNames());
        }}
      />
      <Loading {...appNamesState} />
    </div>
  );
}

function Loading({ type }: ReturnType<typeof selectAppNamesState>) {
  switch (type) {
    case 'reloading': {
      return <LoadingSpinner />;
    }

    default: {
      return null;
    }
  }
}

const mapStateToProps = (state) => ({
  ...state.root,
});

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(
    {
      setQuery,
    },
    dispatch
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(NameSelector);
