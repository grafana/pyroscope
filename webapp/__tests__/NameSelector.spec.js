import React from "react";
import renderer from "react-test-renderer";

import { configure, shallow, mount, render } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import Adapter from "enzyme-adapter-react-16";

import NameSelector from "../javascript/components/NameSelector";

const mockStore = configureMockStore();
configure({ adapter: new Adapter() });

const store = mockStore({});

const props = {
  query: "hotrod.golang.customer",
  names: ["hotrod.golang.customer", "hotrod.golang.driver"],
};

describe("NameSelector", () => {
  it("render correctly NameSelector component", () => {
    const NameSelectorComponent = renderer
      .create(
        <Provider store={store}>
          <NameSelector {...props} />
        </Provider>
      )
      .toJSON();
    expect(NameSelectorComponent).toMatchSnapshot();
  });
  it("When the name in the Applications dropdown menu changes, the page re-renders", () => {
    const wrapper = mount(
      <Provider store={store}>
        <NameSelector {...props} />
      </Provider>
    );

    wrapper.find("select").simulate("mouseDown");
    expect(wrapper.find("option").length).toEqual(3);
    wrapper.find("option").at(2).simulate("click", null);
  });
});
