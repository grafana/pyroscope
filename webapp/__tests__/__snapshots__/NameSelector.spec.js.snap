// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NameSelector render correctly NameSelector component 1`] = `
<span>
  Application:Â 
  <select
    className="label-select"
    onChange={[Function]}
    value="hotrod.golang.customer"
  >
    <option
      disabled={true}
      value="Select an app..."
    >
      Select an app...
    </option>
    <option
      value="hotrod.golang.customer"
    >
      hotrod.golang.customer
    </option>
    <option
      value="hotrod.golang.driver"
    >
      hotrod.golang.driver
    </option>
  </select>
</span>
`;
